optuna/integration/botorch.py

```diff
@@ -0,0 +1,468 @@
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import Sequence
+from typing import Tuple
+
+import numpy
+
+from optuna import multi_objective
+from optuna._experimental import experimental
+from optuna._imports import try_import
+from optuna._transform import _SearchSpaceTransform
+from optuna.distributions import BaseDistribution
+from optuna.distributions import LogUniformDistribution
+from optuna.distributions import UniformDistribution
+from optuna.multi_objective.samplers import BaseMultiObjectiveSampler
+from optuna.multi_objective.samplers._random import RandomMultiObjectiveSampler
+from optuna.multi_objective.study import MultiObjectiveStudy
+from optuna.multi_objective.trial import FrozenMultiObjectiveTrial
+from optuna.samplers import IntersectionSearchSpace
+from optuna.study import StudyDirection
+from optuna.trial import TrialState
+
+
+with try_import() as _imports:
+    from botorch.acquisition.monte_carlo import qExpectedImprovement
+    from botorch.acquisition.multi_objective.monte_carlo import qExpectedHypervolumeImprovement
+    from botorch.acquisition.multi_objective.objective import IdentityMCMultiOutputObjective
+    from botorch.acquisition.objective import ConstrainedMCObjective
+    from botorch.fit import fit_gpytorch_model
+    from botorch.models import SingleTaskGP
+    from botorch.models.transforms.outcome import Standardize
+    from botorch.optim import optimize_acqf
+    from botorch.sampling.samplers import SobolQMCNormalSampler
+    from botorch.utils.multi_objective.box_decomposition import NondominatedPartitioning
+    from botorch.utils.transforms import normalize
+    from botorch.utils.transforms import unnormalize
+    from gpytorch.mlls import ExactMarginalLogLikelihood
+    import torch
+
+
+@experimental("2.4.0")
+def qei_candidates_func(
+    train_x: "torch.Tensor",
+    train_obj: "torch.Tensor",
+    train_con: Optional["torch.Tensor"],
+    bounds: "torch.Tensor",
+) -> "torch.Tensor":
+    """Quasi MC-based batch Expected Improvement (qEI).
+
+    The default value of ``candidates_func`` in :class:`~optuna.integration.BoTorchSampler`
+    with single-objective optimization.
+
+    Args:
+        train_x:
+            Previous parameter configurations. A ``torch.Tensor`` of shape
+            ``(n_trials, n_params)``. ``n_trials`` is the number of already observed trials
+            and ``n_params`` is the number of parameters. ``n_params`` may be larger than the
+            actual number of parameters if categorical parameters are included in the search
+            space, since these parameters are one-hot encoded.
+            Values are not normalized.
+        train_obj:
+            Previously observed objectives. A ``torch.Tensor`` of shape
+            ``(n_trials, n_objectives)``. ``n_trials`` is identical to that of ``train_x``.
+            ``n_objectives`` is the number of objectives. Observations are not normalized.
```

Comment:
They are not normalized when given to the callback, it's up the implementation of the callback to normalize/standardize. We do this for the default provided callbacks. I think this is better since one familiar with BoTorch enough to implement his/her own callback should be aware of how to do this, and this documentation just aids along the way.

Reply:
I understood. Let me resolve this conversation.